File: src/python/xerxes/runtime/config_context.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Process-global runtime config snapshot and event bus.

Stores the active CLI/daemon configuration as a lock-protected dict that
modules (sub-agent spawners, tool wrappers, telemetry) can consult without
plumbing it through every call. :data:`_INHERITABLE` lists the keys that
sub-agents should pick up from their parent. A single optional event callback
funnels lightweight events (sub-agent start/stop, model change, etc.) into
whichever bridge or telemetry sink the host installs.
"""

from __future__ import annotations

import threading
from collections.abc import Callable
from typing import Any

_lock = threading.Lock()
_config: dict[str, Any] = {}
_event_callback: Callable[[str, dict[str, Any]], None] | None = None

_INHERITABLE = {
    "model",
    "base_url",
    "api_key",
    "temperature",
    "top_p",
    "top_k",
    "max_tokens",
    "min_p",
    "frequency_penalty",
    "presence_penalty",
    "repetition_penalty",
    "permission_mode",
}


def set_config(config: dict[str, Any]) -> None:
    """Replace the process-global config snapshot with a copy of ``config``."""

    global _config
    with _lock:
        _config = dict(config)


def get_config() -> dict[str, Any]:
    """Return a shallow copy of the current global config."""

    with _lock:
        return dict(_config)


def get_inheritable() -> dict[str, Any]:
    """Return only the config keys that sub-agents should inherit.

    Filters :func:`get_config` to the whitelist in :data:`_INHERITABLE`,
    skipping ``None`` and empty-string values so callers can ``**spread``
    the result without clobbering child defaults.
    """

    with _lock:
        return {k: v for k, v in _config.items() if k in _INHERITABLE and v is not None and v != ""}


def set_event_callback(cb: Callable[[str, dict[str, Any]], None] | None) -> None:
    """Install (or clear with ``None``) the global event-bus callback."""

    global _event_callback
    with _lock:
        _event_callback = cb


def get_event_callback() -> Callable[[str, dict[str, Any]], None] | None:
    """Return the currently installed event callback, or ``None``."""

    with _lock:
        return _event_callback


def emit_event(event_type: str, data: dict[str, Any]) -> None:
    """Forward ``(event_type, data)`` to the installed callback, swallowing errors.

    Exceptions raised by the callback are intentionally suppressed so a
    misbehaving telemetry sink can't crash the agent loop.
    """

    cb = get_event_callback()
    if cb is not None:
        try:
            cb(event_type, data)
        except Exception:
            pass

File: xerxes/src/runtime/configContext.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { AsyncLocalStorage } from 'node:async_hooks'

export const INHERITABLE_CONFIG_KEYS = [
  'model',
  'base_url',
  'api_key',
  'temperature',
  'top_p',
  'top_k',
  'max_tokens',
  'min_p',
  'frequency_penalty',
  'presence_penalty',
  'repetition_penalty',
  'permission_mode',
] as const

export type RuntimeConfig = Readonly<Record<string, unknown>>
export type RuntimeEventCallback = (eventType: string, data: RuntimeConfig) => void

const inheritableKeys = new Set<string>(INHERITABLE_CONFIG_KEYS)
const activeConfigStorage = new AsyncLocalStorage<RuntimeConfig | undefined>()

let globalConfig: RuntimeConfig = {}
let eventCallback: RuntimeEventCallback | undefined

/** Replace the process-wide configuration snapshot with a defensive copy. */
export function setConfig(config: Readonly<Record<string, unknown>>): void {
  globalConfig = { ...config }
}

/** Return a fresh snapshot so consumers cannot mutate process-global state. */
export function getConfig(): Record<string, unknown> {
  return { ...globalConfig }
}

/**
 * Install the active turn configuration for the current async execution chain.
 *
 * New code should favor runWithActiveConfig because it restores context
 * automatically. This convenience is retained for synchronous daemon setup.
 */
export function setActiveConfig(config: Readonly<Record<string, unknown>> | undefined): void {
  activeConfigStorage.enterWith(config === undefined ? undefined : { ...config })
}

/** Return a copy of the configuration bound to this async turn, if any. */
export function getActiveConfig(): Record<string, unknown> | undefined {
  const value = activeConfigStorage.getStore()
  return value === undefined ? undefined : { ...value }
}

/** Execute work with an isolated active turn configuration. */
export function runWithActiveConfig<T>(
  config: Readonly<Record<string, unknown>> | undefined,
  operation: () => T,
): T {
  return activeConfigStorage.run(config === undefined ? undefined : { ...config }, operation)
}

/** Select non-empty values the spawned-agent surface is allowed to inherit. */
export function getInheritableConfig(): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(globalConfig).filter(([key, value]) => (
      inheritableKeys.has(key) && value !== undefined && value !== null && value !== ''
    )),
  )
}

/** Install or remove the optional process-level runtime event sink. */
export function setEventCallback(callback: RuntimeEventCallback | undefined): void {
  eventCallback = callback
}

/** Return the installed runtime event sink, if any. */
export function getEventCallback(): RuntimeEventCallback | undefined {
  return eventCallback
}

/**
 * Publish a lightweight event without allowing observer failures to break a turn.
 *
 * The callback boundary is intentionally isolated because telemetry and bridge
 * observers are not authorized to alter execution control flow.
 */
export function emitRuntimeEvent(eventType: string, data: Readonly<Record<string, unknown>>): void {
  const callback = eventCallback
  if (callback === undefined) return
  try {
    callback(eventType, { ...data })
  } catch {
    // Observability callbacks must never terminate a model/tool turn.
  }
}
